syntax = "proto3";

package examples.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/any.proto";
import "google/protobuf/struct.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/wrappers.proto";

option go_package = "examples/gen/wellknown/v1;wellknownv1";

// Request with well-known types
message WellKnownTypesRequest {
  // Timestamp field
  google.protobuf.Timestamp created_at = 1;
  
  // Duration field
  google.protobuf.Duration timeout = 2;
  
  // Any field
  google.protobuf.Any metadata = 3;
  
  // Struct field
  google.protobuf.Struct config = 4;
  
  // Wrapper types
  google.protobuf.StringValue title = 5;
  google.protobuf.Int32Value count = 6;
  google.protobuf.BoolValue enabled = 7;
  
  // Optional well-known types
  optional google.protobuf.Timestamp updated_at = 8;
  optional google.protobuf.Duration retry_interval = 9;
}

message WellKnownTypesResponse {
  google.protobuf.Timestamp processed_at = 1;
  google.protobuf.Struct result = 2;
  bool success = 3;
}

// Service for testing well-known types
service WellKnownService {
  rpc ProcessWithWellKnownTypes(WellKnownTypesRequest) returns (WellKnownTypesResponse) {}
  rpc GetStatus(google.protobuf.Empty) returns (WellKnownTypesResponse) {}
}